### PR TITLE
[Feature] Switch between different LLM, Programming Language & UI Interaction

### LINKED ISSUE #10: Design and implement a responsive and friendly user interface for LLM switching functions.
This App should incorporate a user interface for users to smoothly switch between different LLM APIs.

## Tasks

- Implement a function to allow users to switch between different LLM APIs.
- Design a a responsive and friendly user interface for a convenient manipulation.
- Ensure the actions on user interface correspond to the correct and effective functions.

### PR DESCRIPTION
**Pull request by RyanbowZ**
This PR completed the functions of switch between different LLMs, and programming language (C/C++, Java, Javascript, Typescript, Python). Meanwhile, it builds an interactive user interface allows users to directly manipulate on it to make those operations.

## Issue(s)

Issue #9 is closed with the following available LLM list given by Mr. Sing.Li in Rocket.Chat channel:

- Mistral_7b
- Llama3_70b
- codellama-7b
- codestral-22b

You can also use slash command `/ai-programmer list` to view the list within the App.

This PR's function updates also closed #10 #11 #12 #13.

## Acceptance Criteria fulfillment

- [x] Enable the invoking of App directly in the RC channel chat box rather than in separate threads.
- [x] Users can switch between different LLMs, select their expected coding language.
- [x] Implemented the user interface (contextual bar) for users to directly operate functions on.
- [x] Added two LLM options (codellama-7b, codestral-22b) for users to choose from.
- [x] Clarified the scope of available LLM options with help from Sing Li and listed them inside the app. 
- [x] Organized the mechanisms for commands, params, persistence variables, handlers in code.

## Proposed Changes

This demo is prepared for users to have an initial impression about the final product.

It has established a user interface using ui-kit with functions allowing for LLM switch and programming language selection.

Specifically, the App currently allows users to conduct the following four operations with corresponding slash commands:
1. Generate code pieces with specific description -> `/ai-programmer gen code_content`
2. Set the language you want to use to generate code -> `/ai-programmer set C++`
3. Switch to the LLM you want to use to generate code -> `/ai-programmer llm llama3-70b`
4. Use the interactive user interface to handle your operations -> `/ai-programmer ui `
5. List the available LLM options -> `/ai-programmer list `

## Results

image

https://github.com/RocketChat/Apps.RC.AI.Programmer/assets/63179667/d54c0333-370d-4486-b20f-a3d9a074828d

### CODE REVIEW

File: AiProgrammerApp.ts [hunk 1]
```diff
@@ -1,18 +1,32 @@
...
+import { IUIKitResponse, UIKitActionButtonInteractionContext, UIKitBlockInteractionContext, UIKitViewCloseInteractionContext, UIKitViewSubmitInteractionContext } from '@rocket.chat/apps-engine/definition/uikit';
+import { ElementBuilder } from "./lib/ElementBuilder";
+import { BlockBuilder } from "./lib/BlockBuilder";
+import { ExecuteActionButtonHandler } from './handlers/ExecuteActionButtonHandler';
+import { ExecuteBlockActionHandler } from './handlers/ExecuteBlockActionHandler';
+import { ExecuteViewClosedHandler } from './handlers/ExecuteViewClosedHandler';
+import { ExecuteViewSubmitHandler } from './handlers/ExecuteViewSubmitHandler';
 
 export class AiProgrammerApp extends App {
+	private elementBuilder: ElementBuilder;
```
Comment (samad-yar-khan) [inline AiProgrammerApp.ts:23]: nit:fix indentation

File: constants/CodePrompts.ts [hunk 2]
```diff
@@ -1,7 +1,6 @@
-const CODE_PROMPT = `You are a system assistant that helps users write short code modules 
-in C/C++, Java, Javascript, Typescript or Python based on the specification supplied by the user, the specifications are separated by double slashes as (//): {dialogue}. 
-Please generate the code and in the first row of your response describe the main funtionality of this code pieces in less than 20 words. Remember to wrap your code piece part with single quote pairs.`;
-
-export function generateCodePrompt(dialogue: string): string {
-	return CODE_PROMPT.replace('{dialogue}', dialogue);
+const CODE_PROMPT = `You are a system assistant that helps users generate short code modules 
+in {language} language based on the specification supplied by the user: {dialogue}. 
+Please generate the code with proper format.`;
```
Comment (samad-yar-khan) [inline constants/CodePrompts.ts:3]: We need to work on improving this  prompt to make sure that the user is not able to bypass this beyond a code gen app

File: helpers/generateCode.ts [hunk 3]
```diff
@@ -1,23 +1,30 @@
...
+    language: string,
+    LLM: string,
+	prompt: string
 ): Promise<string> {
 	const model = await app
 		.getAccessors()
 		.environmentReader.getSettings()
 		.getValueById('model');
-	const url = `http://${model}/v1`;
+	//const url = `http://${model}/v1`;
```
Comment (samad-yar-khan) [inline helpers/generateCode.ts:26]: remove comment

File: helpers/generateCode.ts [hunk 4]
```diff
@@ -38,13 +45,13 @@ export async function createTextCompletion(
...
-			'Something is wrong with AI. Please try again later',
-			threadId
-		);
+		await sendNotification(
+            read,
+            modify,
+            user,
+            room,
+            "Something is wrong with AI. Please try again later! "
+        );
```
Comment (samad-yar-khan) [inline helpers/generateCode.ts:54]: fix indentation

Comment (samad-yar-khan) [reply helpers/generateCode.ts:54]: would be a good idea to use a linter

**Review by samad-yar-khan — CHANGES_REQUESTED** [review]

**Review by samad-yar-khan — APPROVED** [review]
@RyanbowZ Please cater to these comments in the next PR as this PR is already big enough. I am merging this for now

**Comment by mustafahasankhan** [comment]
LGTM

**Pull request merged**